modules/apps/transfer/types/v3/packet.go

```diff
@@ -13,3 +34,103 @@ func NewFungibleTokenPacketData(
 		Memo:     memo,
 	}
 }
+
+// ValidateBasic is used for validating the token transfer.
+// NOTE: The addresses formats are not validated as the sender and recipient can have different
+// formats defined by their corresponding chains that are not known to IBC.
+func (ftpd FungibleTokenPacketData) ValidateBasic() error {
+	if strings.TrimSpace(ftpd.Sender) == "" {
+		return errorsmod.Wrap(ibcerrors.ErrInvalidAddress, "sender address cannot be blank")
+	}
+
+	if strings.TrimSpace(ftpd.Receiver) == "" {
+		return errorsmod.Wrap(ibcerrors.ErrInvalidAddress, "receiver address cannot be blank")
+	}
+
+	if len(ftpd.Tokens) == 0 {
+		return errorsmod.Wrap(types.ErrInvalidAmount, "tokens cannot be empty")
+	}
+
+	for _, token := range ftpd.Tokens {
+		amount, ok := sdkmath.NewIntFromString(token.Amount)
+		if !ok {
+			return errorsmod.Wrapf(types.ErrInvalidAmount, "unable to parse transfer amount (%s) into math.Int", token.Amount)
+		}
+
+		if !amount.IsPositive() {
+			return errorsmod.Wrapf(types.ErrInvalidAmount, "amount must be strictly positive: got %d", amount)
+		}
+
+		if err := token.Validate(); err != nil {
+			return err
+		}
+	}
+
+	if len(ftpd.Memo) > types.MaximumMemoLength {
+		return errorsmod.Wrapf(types.ErrInvalidMemo, "memo must not exceed %d bytes", types.MaximumMemoLength)
+	}
+
+	return nil
+}
+
+// ValidateToken validates a token denomination and trace identifiers.
+func (t *Token) Validate() error {
+	if err := sdk.ValidateDenom(t.Denom); err != nil {
+		return errorsmod.Wrap(types.ErrInvalidDenomForTransfer, err.Error())
+	}
+
+	trace := strings.Join(t.Trace, "/")
+	identifiers := strings.Split(trace, "/")
```

Comment:
I'd rather avoid modifying `ValidateTraceIdentifiers` since that is being used by both `DenomTrace.Validate()` and `ValidatePrefixedDenom()` on the fungible token packet data v1, and the idea is that both v2 and v1 could use this function. I can check out how to implement using slices.